README.md
# TrustedGRUB2

[![Build Status](https://travis-ci.org/Rohde-Schwarz-Cybersecurity/TrustedGRUB2.svg?branch=master)](https://travis-ci.org/Rohde-Schwarz-Cybersecurity/TrustedGRUB2)

<a href="https://scan.coverity.com/projects/rohde-schwarz-cybersecurity-trustedgrub2">
  <img alt="Coverity Scan Build Status"
       src="https://scan.coverity.com/projects/5521/badge.svg"/>
</a>

[![Quality Gate](https://sonarqube.com/api/badges/gate?key=TrustedGRUB2)](https://sonarqube.com/dashboard/index/TrustedGRUB2)

## 1. General Information

### 1.1 Introduction

This file describes the extensions made to transform a standard GRUB2 into a version that offers TCG (TPM) support for granting the integrity of the boot process (trusted boot). This project was highly inspired by the former projects [TrustedGrub1](https://www.sirrix.com/content/pages/trustedgrub_en.htm) and GRUB-IMA. However TrustedGRUB2 was completely written from scratch.

TrustedGRUB2 is measuring all critical components during the boot process, i.e. GRUB2 kernel, GRUB2 modules, the OS kernel or OS modules and so on, together with their
parameters. Please note that the TrustedGRUB2 MBR bootcode has not to be checked here (it wouldn't even be possible). The MBR bootcode has already been measured by the TPM itself.
Since the TPM is passive, it has no direct ability to check if the integrity of bootloader (and the OS kernel/modules and so on) actually is correct.
This can only be done indirectly by using the seal/unseal functions of the TPM (for details on this topic, you should have a look at the TCG specifications or on other documents describing TCG/TPM abilities).

### 1.2 Features

* Based on GRUB 2.02
* TPM Support with TPM detection (only legacy/mbr mode, UEFI is not supported at the moment)
* Measurement of GRUB2 kernel
* Measurement of all loaded GRUB2 modules
* Measurement of all commands and their parameters entered in shell and scripts
* New SHA1-implementation in GRUB2 kernel (necessary for doing the GRUB2 modules measurement as the crypto module isn't loaded at this stage)
* Added LUKS keyfile support with additional parameter `-k KEYFILE` for `cryptomount` command
* Added support for unsealing LUKS keyfile with additional `-s` parameter for `cryptomount` command. LUKS-header is measured before unsealing into PCR 12. Currently unsealing only supported with SRK and well known secret (20 zero bytes)
* New commands:
  * `readpcr PCRNUM`
  * `tcglog LOGINDEX`
  * `measure FILE PCRNUM`
  * `setmor DISABLEAUTODETECT`
* Loader measurements:
  * `linux` / `linux16`
  * `initrd` / `initrd16`
  * `chainloader`
  * `ntdlr`
  * `multiboot`
  * `module`
* New cryptomount parameters:
  * `cryptomount -k KEYFILE`
  * `cryptomount -k KEYFILE -s`
* Functionality added without own command:
  * TPM_Unseal
  * TPM_GetRandom
  * TPM_OIAP
  * TPM_OSAP

### 1.3 Measurements (in short)

* PCR 0-7 Measured by BIOS
* PCR 8 First sector of TrustedGRUB2 kernel (diskboot.img)
* PCR 9 TrustedGRUB2 kernel (core.img)
* PCR 10 Loader measurements - currently linux-kernel, initrd, ntldr, chainloader, multiboot, module
* PCR 11 Contains all commandline arguments from scripts (e.g. grub.cfg) and those entered in the shell
* PCR 12 LUKS-header
* PCR 13 Parts of GRUB2 that are loaded from disk like GRUB2-modules // TODO: fonts, themes, locales

Kernel measurements are only implemented for diskboot so far (e.g. no cdboot or pxeboot measurement)

### 1.4 Requirements

In order to use the TCG-enhanced TrustedGRUB2, you need a computer which has TCG enhancements according to TCG specs. v1.2, since SHA1-calculations are extended into PC-Registers of the TPM.

### 1.5 Known Bugs / Limitations

* On some HP notebooks and workstations, TrustedGRUB2 (in default mode) is not able to do the kernel measurements due to a buggy BIOS. This means PCR 9 can 
contain bogus values. HP desktop/laptop BIOS seems to be unable to handle blocks ending on 512 byte boundaries when measuring data.
  * Fortunately we've found a workaround:
    * The workaround works as follows: we increase the number of bytes to read by 1 and also the number of sectors to read, which ensures that all 
bytes of core.img are read. For this to work correctly the loaded core.img must be padded with zeroes.
    * In summary: 
      1. pad core.img with zeroes to 512 byte blocks.
      2. append 1 extra zero byte to core.img.
    * This doesn't have to be done manually. We've patched `grub_mkimage` to do step 1 and step 2 for us.
    * This workaround has to be enabled explicitly. To do so: define `TGRUB_HP_WORKAROUND`. For example like this: `make 
CPPFLAGS=-DTGRUB_HP_WORKAROUND`
    * IMPORTANT: you have to append `--no-rs-codes` to `grub-install` otherwise you end up in a reboot loop.

If you find any other bugs, create an issue on github

### 1.6 Configuring TrustedGRUB2 before installation

#### 1.6.1 PCR selection
PCR selection for module measurement, command measurement and loaded files measurement can be adjusted in tpm.h:

```C++
#define TPM_LOADER_MEASUREMENT_PCR 10
#define TPM_COMMAND_MEASUREMENT_PCR 11
#define TPM_LUKS_HEADER_MEASUREMENT_PCR 12
#define TPM_GRUB2_LOADED_FILES_MEASUREMENT_PCR 13
```

#### 1.6.2 Debug output

To enable some debug output define `TGRUB_DEBUG`. For example like this `make CPPFLAGS=-DTGRUB_DEBUG`

### 1.7 Installation of TrustedGRUB2

Required Packages for compiling:
* autogen
* autoconf
* automake
* gcc
* bison
* flex

To compile and install TrustedGRUB2, please run

```bash
./autogen.sh
./configure --prefix=INSTALLDIR --target=i386 -with-platform=pc
make
make install
```

Installing to device:

```bash
./INSTALLDIR/sbin/grub-install --directory=INSTALLDIR/lib/grub/i386-pc /dev/sda 
```

[WARNING]
if installing over an old GRUB2 install you probably have to adjust your grub.cfg

For usb-devices this command can be used (assuming /dev/sdb/ is your usb-device):

```bash
./INSTALLDIR/sbin/grub-install --directory=INSTALLDIR/lib/grub/i386-pc --root-directory=/mnt/sdb1 /dev/sdb
```

## 2. Technical Details

### 2.1 General view on how TrustedGRUB2 works

The goal of TrustedGRUB2 is to accomplish a chain of trust, i.e. every component measures the integrity of the succeeding component.
Concretely, this looks like the following:

|         Component							   |		measured by              |
| -------------------------------------------  |  ----------------------------- |
| BIOS                  					   | CRTM					|
| TrustedGRUB2 MBR bootcode    				   | BIOS					|
| start of TrustedGRUB2 kernel (diskboot.img)  | TrustedGRUB2 MBR bootcode		|
| rest of TrustedGRUB2 kernel (core.img)	   | start of TrustedGRUB2 kernel	|
| Grub modules + OS (kernel and so on) 		   | TrustedGRUB2 kernel			|

This chain of trust can be extended by using the newly added `measure` command to measure the integrity of arbitrary files.

### 2.2 Measurement of GRUB2 kernel

#### 2.2.1 Modifications in boot.S (MBR bootcode)

GRUB2 MBR bootcode is already measured by the TPM. The MBR bootcode has the task to load first sector of TrustedGRUB2 kernel (diskboot.img). Diskboot.img itself loads the rest of GRUB2 kernel.
Therefore GRUB2 MBR code is extended to measure diskboot.img before jumping to it:

1. Diskboot.img is hashed with a SHA-1 algorithm. Diskboot.img is loaded at address 0x8000, its length is 512 bytes.
2. The resulting hash value is written to PCR (Platform Configuration Register) 8. More precisely, the former content of this register (which actually is 0) is concatenated to the new value, then hashed with SHA1 and finally written again to PCR 8

Due to the PC architecture, the size of the MBR (where TrustedGRUB2 boot.S is
located) is limited to 512 bytes. But the original GRUB2 MBR bootcode is already very
close to this limit, leaving very few space for the TCG extensions. Because
of this, it was necessary (in the current version of TrustedGRUB2) to eliminate the CHS-code.
This results in the problem that we support only LBA-discs now. FDD boot is not possible.

#### 2.2.2 Modifications in diskboot.S

boot.S contains the code for loading the first sector of TrustedGRUB2 kernel (diskboot.img). Its only task
is the load the rest of TrustedGRUB2 kernel. Therefore, the TCG extension now has to measure the rest of TrustedGRUB2 kernel
The changes here are widely the same as in TrustedGRUB2 bootcode, with the differences that
the entry point for the code which has to be checked is a address 0x8200 and that the result is written into PCR 9.

### 2.3 Measurement of GRUB2 modules

Grub2 has a modular structure. GRUB2 dynamically loads needed modules which are not contained in kernel. Modifications in boot.S and diskboot.S are only measuring GRUB2 kernel.
Therefore the GRUB2 module loader was modified to measure modules to PCR 13 before they are loaded. Changes can be found in dl.c .

### 2.4 New SHA1-implementation in GRUB2 kernel

In order to make GRUB2 modules measurement possible, a SHA1-implementation had to be added to the kernel.
GRUB2 already contains an SHA1-implementation in its crypto module, but this isn't loaded at this stage.

### 2.5 Measurement of all commands and their parameters entered in shell and scripts

All commands which are entered in shell or executed by scripts is measured to PCR 11. Therefore commands in grub.cfg are automatically measured. No need to measure grub.cfg separately.
One exception applies to this rule: The `menuentry`, `submenu` and `[ ... ]` commands are not measured because it makes precomputation of the PCR 
value difficult and is unnecessary because each command within `menuentry` or `submenu` is anyway measured. For `[ ... ]` it shouldn't be possible to 
write commands between the square brackets.

### 2.6 TrustedGRUB2 commands

#### 2.6.1 new commands

```
readpcr PCRNUM
```

Display current value of the PCR (Platform Configuration Register) within TPM (Trusted Platform Module) at index, `PCRNUM`.  
<br>  
<br>  

```
tcglog LOGINDEX
```

Displays TCG event log entry at position, `LOGINDEX`. Type in "0" for all entries.  
<br>  
<br>  

```
measure FILE PCRNUM
```

Perform TCG measurement operation with the file `FILE` and with PCR( `PCRNUM` ).  
<br>  
<br>  

```
setmor DISABLEAUTODETECT
```

Sets Memory Overwrite Request (MOR) Bit. `DISABLEAUTODETECT` specifies if BIOS should auto detect unscheduled reboots.  
<br>  
<br>  

#### 2.6.2 Modified existing GRUB2 commands

* `linux` / `linux16`  
* `initrd` / `initrd16`  
* `chainloader`  
* `ntdlr`
* `multiboot`
* `module`
  * append `--nounzip` to get measuremens of the compressed file 

These commands are modified to measure before loading. PCR 10 is extended.

Additionally the following commands have been modified:

* `cryptomount`

### 2.7 Other modifications

All modifications have been commented with

```C++
/* BEGIN TCG EXTENSION */

/* END TCG EXTENSION */
```

### 2.8 Security considerations

* The `multiboot` command measurement does not follow the new convention of measuring the same buffer that is loaded into memory. If someone needs this extra security feel free to send a pull request. See GH #9 and GH #38 for more details.

### 2.9 File list

The following list presents the files that have been added / modified to add TCG
support to GRUB2.

* README.md
* Changelog.md
* grub-core/Makefile.am
* grub-core/Makefile.core.def
* grub-core/boot/i386/pc/boot.S
* grub-core/boot/i386/pc/diskboot.S
* grub-core/kern/tpm.c
* grub-core/kern/dl.c
* grub-core/kern/i386/pc/tpm/tpm_kern.c
* grub-core/kern/sha1.c
* grub-core/disk/cryptodisk.c
* grub-core/disk/luks.c
* grub-core/loader/multiboot.c
* grub-core/loader/linux.c
* grub-core/loader/i386/linux.c
* grub-core/loader/i386/pc/chainloader.c
* grub-core/loader/i386/pc/linux.c
* grub-core/loader/i386/pc/ntldr.c
* grub-core/normal/main.c
* grub-core/script/execute.c
* grub-core/tpm/i386/pc/tpm.c
* include/grub/i386/pc/boot.h
* include/grub/i386/pc/tpm.h
* include/grub/tpm.h
* include/grub/sha1.h
* util/mkimage.c

## 3. Thanks

[TrustedGrub1](https://www.sirrix.com/content/pages/trustedgrub_en.htm) and GRUB-IMA have done a lot of preparatory work in the field and were used for code examples.


Changelog.md
### Changelog

#### 1.5.0
* Update to GRUB 2.02

#### 1.4.0
* Extend log on each TPM_Extend GH #11

#### 1.3.0

* Fix measurement of compressed files. Previously the uncompressed version of the file was measured and not the one that is stored on the disk. GH #28
* Command measurement: in addition to not measuring `menuentry` also `submenu` and `[ ... ]` are not measured to simplify precomputation. GH #25
* Update to latest GRUB2 master (23.12.2015) that also includes a fix for CVE-2015-8370
* Prevent possible buffer overlow in case the command to measure is greater than 1024 byte in length
* Disable HP workaround in default mode, i.e. HP workaround has to be enabled explicitly by defining `TGRUB_HP_WORKAROUND`. Therefore there is no need to append `--no-rs-codes` to `grub-install` anymore in case you don't need the workaround. GH #18
* Measure buffer that is used. Before this fix everything that was measured from disk was read a second time. This enabled following attack: A sufficiently malicious storage device might provide a backdoored file on the first read attempt, followed by the correct file on the second read attempt. The measurement would then appear correct. GH #9
* Measurements of parts of TrustedGRUB2 that are loaded at runtime like grub2-modules are now seperated from the loader measurements like kernel and initrd. Additionally renamed `TPM_LOADED_FILES_PCR` to `TPM_LOADER_MEASUREMENT_PCR` and introduced a new define `TPM_GRUB2_LOADED_FILES_MEASUREMENT_PCR` for the GRUB2 measurements. GH #34
* Add multiboot measurements for `multiboot` and `module` commands. For now the `multiboot` measurement does not follow the new convention of measuring the same buffer that is loaded into memory. If someone needs this extra security feel free to send a pull request. GH #35

#### 1.2.1

* Implemented a workaround for buggy HP desktop/laptop BIOS. 
  * Measurement of PCR 8 / PCR 9 should be correct now on these devices.The way the core.img is created / measured has changed a bit. So resealing is necessary.
  * For the moment it is necessary to call `grub-install` with `--no-rs-codes`

#### 1.2.0

* sync with upstream GRUB2
* use build in byte swap functions instead of own
* improved debug output
* some general cleanup

#### 1.1.0

* fixed wrong parameter type sizes in grub_TPM_openOSAP_Session
* use grub_bios_interrupt() instead of self written assembler functions
* check if TPM is present but deactivated

#### 1.0.3

* fix out-of-bounds read in cryptomount command

#### 1.0.2

* abort boot process in case of errors during TrustedGRUB2 kernel measurement

#### 1.0.1

* treat errors in tpm functionality as fatal erros
* fixed bug in cryptomount command, if there is more than one keyslot

#### 1.0.0

* initial release


grub-core/kern/sha1.c
/* Begin TCG Extension */

/*  This implementation is needed because we have to measure some parts before the crypto module is loaded */

/*  This SHA1-implementation has been written by Marko Wolf <mwolf@crypto.rub.de> and tested according to FIPS-180.
	The SHA1-macros are from "Christophe Devine" <devine@cr0.net>.

	Parameters:
    int sha1_init(sha1_context *ctx )
	int sha1_update(sha1_context *ctx, t_U8 *chunk_data, t_U32 chunk_length)
	int sha1_finish(sha1_context *ctx, t_U32 *sha1_hash)
*/

/* Added by TrustedGRUB2 author:
	sha1_hash_file( const grub_file_t file, void* result )
	sha1_hash_string( const char* string, void* result )
*/

#include <grub/sha1.h>
#include <grub/types.h>
#include <grub/mm.h>
#include <grub/misc.h>

// concatenates 4 * 8-bit words (= 1 byte) to one 32-bit word
#define CONCAT_4_BYTES( w32, w8, w8_i)            \
{                                                 \
    (w32) = ( (t_U32) (w8)[(w8_i)    ] << 24 ) |  \
            ( (t_U32) (w8)[(w8_i) + 1] << 16 ) |  \
            ( (t_U32) (w8)[(w8_i) + 2] <<  8 ) |  \
            ( (t_U32) (w8)[(w8_i) + 3]       );   \
}

// splits a 32-bit word into 4 * 8-bit words (= 1 byte)
#define SPLIT_INTO_4_BYTES( w32, w8, w8_i)        \
{                                                 \
    (w8)[(w8_i)    ] = (t_U8) ( (w32) >> 24 );    \
    (w8)[(w8_i) + 1] = (t_U8) ( (w32) >> 16 );    \
    (w8)[(w8_i) + 2] = (t_U8) ( (w32) >>  8 );    \
    (w8)[(w8_i) + 3] = (t_U8) ( (w32)       );    \
}

// FIPS-180-1 padding sequence
static t_U8 sha1_padding[64] =
{
 (t_U8) 0x80, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0,
 (t_U8)    0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0,
 (t_U8)    0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0,
 (t_U8)    0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0, (t_U8) 0
};

static int sha1_init(sha1_context *ctx )
{

  // parameter check
  if ( ctx == NULL )
  {
    return -1;
  }

  // byte length = 0
  ctx->total_bytes_Lo = 0;
  ctx->total_bytes_Hi = 0;

  // FIPS 180-1 init values
  ctx->vector[0] = 0x67452301;
  ctx->vector[1] = 0xEFCDAB89;
  ctx->vector[2] = 0x98BADCFE;
  ctx->vector[3] = 0x10325476;
  ctx->vector[4] = 0xC3D2E1F0;

  // successful
  return 0;
}


static void sha1_process(sha1_context *ctx, t_U8 *byte_64_block )
{
  // declarations
  t_U32 temp, W[16];
  t_U32 A, B, C, D, E;

  // concatenate 64 bytes to 16 * 32-bit words
  CONCAT_4_BYTES( W[0],  byte_64_block,  0 );
  CONCAT_4_BYTES( W[1],  byte_64_block,  4 );
  CONCAT_4_BYTES( W[2],  byte_64_block,  8 );
  CONCAT_4_BYTES( W[3],  byte_64_block, 12 );
  CONCAT_4_BYTES( W[4],  byte_64_block, 16 );
  CONCAT_4_BYTES( W[5],  byte_64_block, 20 );
  CONCAT_4_BYTES( W[6],  byte_64_block, 24 );
  CONCAT_4_BYTES( W[7],  byte_64_block, 28 );
  CONCAT_4_BYTES( W[8],  byte_64_block, 32 );
  CONCAT_4_BYTES( W[9],  byte_64_block, 36 );
  CONCAT_4_BYTES( W[10], byte_64_block, 40 );
  CONCAT_4_BYTES( W[11], byte_64_block, 44 );
  CONCAT_4_BYTES( W[12], byte_64_block, 48 );
  CONCAT_4_BYTES( W[13], byte_64_block, 52 );
  CONCAT_4_BYTES( W[14], byte_64_block, 56 );
  CONCAT_4_BYTES( W[15], byte_64_block, 60 );

// rotate left by n bits
#define ROTATE_N_LEFT(x,n) ((x << n) | ((x & 0xFFFFFFFF) >> (32 - n)))

// extends 16 * 32-bit words to 80 * 32-bit words
#define EXTENDED_W(t)                                 \
(                                                     \
  temp = W[(t -  3) & 0x0F] ^ W[(t - 8) & 0x0F] ^     \
         W[(t - 14) & 0x0F] ^ W[ t      & 0x0F],      \
  ( W[t & 0x0F] = ROTATE_N_LEFT(temp,1) )             \
)

// main formula
#define P(a,b,c,d,e,Wi)                               \
{                                                     \
  e += ROTATE_N_LEFT(a,5) + F(b,c,d) + K + Wi;        \
  b  = ROTATE_N_LEFT(b,30);                           \
}

  // init A..E
  A = ctx->vector[0];
  B = ctx->vector[1];
  C = ctx->vector[2];
  D = ctx->vector[3];
  E = ctx->vector[4];

// round I (0..19)
#define F(x,y,z) (z ^ (x & (y ^ z)))
#define K 0x5A827999

  P( A, B, C, D, E, W[0]  );
  P( E, A, B, C, D, W[1]  );
  P( D, E, A, B, C, W[2]  );
  P( C, D, E, A, B, W[3]  );
  P( B, C, D, E, A, W[4]  );
  P( A, B, C, D, E, W[5]  );
  P( E, A, B, C, D, W[6]  );
  P( D, E, A, B, C, W[7]  );
  P( C, D, E, A, B, W[8]  );
  P( B, C, D, E, A, W[9]  );
  P( A, B, C, D, E, W[10] );
  P( E, A, B, C, D, W[11] );
  P( D, E, A, B, C, W[12] );
  P( C, D, E, A, B, W[13] );
  P( B, C, D, E, A, W[14] );
  P( A, B, C, D, E, W[15] );
  P( E, A, B, C, D, EXTENDED_W(16) );
  P( D, E, A, B, C, EXTENDED_W(17) );
  P( C, D, E, A, B, EXTENDED_W(18) );
  P( B, C, D, E, A, EXTENDED_W(19) );

#undef K
#undef F

// round II (20..39)
#define F(x,y,z) (x ^ y ^ z)
#define K 0x6ED9EBA1

  P( A, B, C, D, E, EXTENDED_W(20) );
  P( E, A, B, C, D, EXTENDED_W(21) );
  P( D, E, A, B, C, EXTENDED_W(22) );
  P( C, D, E, A, B, EXTENDED_W(23) );
  P( B, C, D, E, A, EXTENDED_W(24) );
  P( A, B, C, D, E, EXTENDED_W(25) );
  P( E, A, B, C, D, EXTENDED_W(26) );
  P( D, E, A, B, C, EXTENDED_W(27) );
  P( C, D, E, A, B, EXTENDED_W(28) );
  P( B, C, D, E, A, EXTENDED_W(29) );
  P( A, B, C, D, E, EXTENDED_W(30) );
  P( E, A, B, C, D, EXTENDED_W(31) );
  P( D, E, A, B, C, EXTENDED_W(32) );
  P( C, D, E, A, B, EXTENDED_W(33) );
  P( B, C, D, E, A, EXTENDED_W(34) );
  P( A, B, C, D, E, EXTENDED_W(35) );
  P( E, A, B, C, D, EXTENDED_W(36) );
  P( D, E, A, B, C, EXTENDED_W(37) );
  P( C, D, E, A, B, EXTENDED_W(38) );
  P( B, C, D, E, A, EXTENDED_W(39) );

#undef K
#undef F

// round III (40..59)
#define F(x,y,z) ((x & y) | (z & (x | y)))
#define K 0x8F1BBCDC

  P( A, B, C, D, E, EXTENDED_W(40) );
  P( E, A, B, C, D, EXTENDED_W(41) );
  P( D, E, A, B, C, EXTENDED_W(42) );
  P( C, D, E, A, B, EXTENDED_W(43) );
  P( B, C, D, E, A, EXTENDED_W(44) );
  P( A, B, C, D, E, EXTENDED_W(45) );
  P( E, A, B, C, D, EXTENDED_W(46) );
  P( D, E, A, B, C, EXTENDED_W(47) );
  P( C, D, E, A, B, EXTENDED_W(48) );
  P( B, C, D, E, A, EXTENDED_W(49) );
  P( A, B, C, D, E, EXTENDED_W(50) );
  P( E, A, B, C, D, EXTENDED_W(51) );
  P( D, E, A, B, C, EXTENDED_W(52) );
  P( C, D, E, A, B, EXTENDED_W(53) );
  P( B, C, D, E, A, EXTENDED_W(54) );
  P( A, B, C, D, E, EXTENDED_W(55) );
  P( E, A, B, C, D, EXTENDED_W(56) );
  P( D, E, A, B, C, EXTENDED_W(57) );
  P( C, D, E, A, B, EXTENDED_W(58) );
  P( B, C, D, E, A, EXTENDED_W(59) );

#undef K
#undef F

// round IV (60..79)
#define F(x,y,z) (x ^ y ^ z)
#define K 0xCA62C1D6

  P( A, B, C, D, E, EXTENDED_W(60) );
  P( E, A, B, C, D, EXTENDED_W(61) );
  P( D, E, A, B, C, EXTENDED_W(62) );
  P( C, D, E, A, B, EXTENDED_W(63) );
  P( B, C, D, E, A, EXTENDED_W(64) );
  P( A, B, C, D, E, EXTENDED_W(65) );
  P( E, A, B, C, D, EXTENDED_W(66) );
  P( D, E, A, B, C, EXTENDED_W(67) );
  P( C, D, E, A, B, EXTENDED_W(68) );
  P( B, C, D, E, A, EXTENDED_W(69) );
  P( A, B, C, D, E, EXTENDED_W(70) );
  P( E, A, B, C, D, EXTENDED_W(71) );
  P( D, E, A, B, C, EXTENDED_W(72) );
  P( C, D, E, A, B, EXTENDED_W(73) );
  P( B, C, D, E, A, EXTENDED_W(74) );
  P( A, B, C, D, E, EXTENDED_W(75) );
  P( E, A, B, C, D, EXTENDED_W(76) );
  P( D, E, A, B, C, EXTENDED_W(77) );
  P( C, D, E, A, B, EXTENDED_W(78) );
  P( B, C, D, E, A, EXTENDED_W(79) );

#undef K
#undef F

  // assign vectors
  ctx->vector[0] += A;
  ctx->vector[1] += B;
  ctx->vector[2] += C;
  ctx->vector[3] += D;
  ctx->vector[4] += E;
}

static int sha1_update(sha1_context *ctx, t_U8 *chunk_data, t_U32 chunk_length)
{

  // declarations
  t_U32 left, fill;
  t_U32 i;

  // parameter check
  if ( (ctx == NULL) || (chunk_data == NULL) || (chunk_length < 1) )
  {
    return -1;
  }

  // chunk_length = n * 64 byte + left
  left = ctx->total_bytes_Lo & 0x3F;

  // fill bytes remain to 64 byte block
  fill = 64 - left;

  // total = total + chunk_length
  ctx->total_bytes_Lo += chunk_length;
  
  // mask 32 bit
  ctx->total_bytes_Lo &= 0xFFFFFFFF;

  if ( ctx->total_bytes_Lo < chunk_length )
  {
    ctx->total_bytes_Hi++;
  }

  // if we have something in the buffer (left > 0) and 
  // the chunk has enougth data to fill a 64 byte block (chunk_length >= fill)
  if ( (left > 0) && (chunk_length >= fill) )
  {
     // fill buffer with data from new chunk
     for ( i = 0; i < fill; i++ )
     {
        ctx->buffer[left + i] = chunk_data[i];
     }

     // process 64 byte buffer block
     sha1_process( ctx, ctx->buffer );

     // dec chunk_length by fill
     chunk_length -= fill;

     // move data pointer by fill
     chunk_data  += fill;

     // buffer is fully processed
     left = 0;
  }

  // process all remaining 64 byte chunks
  while( chunk_length >= 64 )
  {
     sha1_process( ctx, chunk_data );
     chunk_length -= 64;
     chunk_data  += 64;
  }

  // if final chunk_length between 1..63 byte
  if ( chunk_length > 0 )
  {
     // append remainder to 64 byte into buffer resp. fill the empty buffer
     for ( i = 0; i < chunk_length; i++ )
     {
       ctx->buffer[left + i] = chunk_data[i];
     }
  }

  // successfull
  return 0;
}

static int sha1_finish(sha1_context *ctx, t_U32 *sha1_hash)
{

  // declarations
  t_U32 last, padn;
  t_U32 high, low;
  t_U8  msglen[8];
  /* int   ret; */

  // parameter check
  if ( (ctx == NULL) || (sha1_hash == NULL) )
  {
    return -1;
  }

  // build msglen array[8 * 8-bit] from total[2 * 32-bit] = n * 64 byte
  high = ( ctx->total_bytes_Lo >> 29 ) | ( ctx->total_bytes_Hi <<  3 );
  low  = ( ctx->total_bytes_Lo <<  3 );
  SPLIT_INTO_4_BYTES( high, msglen, 0 );
  SPLIT_INTO_4_BYTES( low,  msglen, 4 ); 

  // total = n * 64 bytes + last
  last = ctx->total_bytes_Lo & 0x3F;

  // number of padding zeros 
  padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );

  // update SHA-1 context with remaining buffer and padding to 64 bytes with bit sequence (1,0,...,0)
  /*ret = */ sha1_update( ctx, sha1_padding, padn );

  // update SHA-1 context with total length
  /* ret = */ sha1_update( ctx, msglen, 8 );

  // assign final hash words
  sha1_hash[0] = ctx->vector[0];
  sha1_hash[1] = ctx->vector[1];
  sha1_hash[2] = ctx->vector[2];
  sha1_hash[3] = ctx->vector[3];
  sha1_hash[4] = ctx->vector[4];

  // successful
  return 0;
}

/* file has to be opened before call

    Returns on failure:
     'GRUB_ERR_TPM'
     'GRUB_ERR_OUT_OF_RANGE'
*/
grub_err_t
sha1_hash_file( const grub_file_t file, void* result ) {
  sha1_context context;
  grub_uint8_t readbuf[4096];

  if( sha1_init( &context ) != 0 ) {
      return grub_error (GRUB_ERR_TPM, N_("sha1_hash_file: hashing failed"));
  }

  while( 1 ) {
      grub_ssize_t r;
      r = grub_file_read( file, readbuf, sizeof( readbuf ) );

      if ( grub_errno ) {
          return grub_errno;
      }

      if( r == 0 ) {
    	  break;
      }
      if( sha1_update( &context, readbuf, (grub_uint32_t) r ) != 0 ) {
          return grub_error (GRUB_ERR_TPM, N_("sha1_hash_file: hashing failed"));
      }
  }
  if( sha1_finish( &context, result ) != 0 ) {
      return grub_error (GRUB_ERR_TPM, N_("sha1_hash_file: hashing failed"));
  }

  return GRUB_ERR_NONE;
}

/* Returns on failure:
     'GRUB_ERR_TPM'
     'GRUB_ERR_OUT_OF_RANGE'
*/
grub_err_t
sha1_hash_buffer( const void* buffer, const grub_uint32_t bufferLen, void* result ) {
  sha1_context context;

  if( sha1_init( &context ) != 0 ) {
      return grub_error (GRUB_ERR_TPM, N_("sha1_hash_buffer: hashing failed"));
  }

  if( sha1_update( &context, (t_U8*) buffer, bufferLen ) != 0 ) {
	  return grub_error (GRUB_ERR_TPM, N_("sha1_hash_buffer: hashing failed"));
  }

  if( sha1_finish( &context, result ) != 0 ) {
      return grub_error (GRUB_ERR_TPM, N_("sha1_hash_buffer: hashing failed"));
  }

  return GRUB_ERR_NONE;
}

/* Returns on failure:
     'GRUB_ERR_TPM'
*/
grub_err_t
sha1_hash_string( const char* string, void* result ) {
	sha1_context context;

	if( sha1_init( &context ) != 0 ) {
        return grub_error (GRUB_ERR_TPM, N_("sha1_hash_string: hashing failed"));
	}

	if( sha1_update( &context, (t_U8*)string, grub_strlen( string ) ) != 0 ) {
        return grub_error (GRUB_ERR_TPM, N_("sha1_hash_string: hashing failed"));
	}

	if( sha1_finish( &context, result ) != 0 ) {
        return grub_error (GRUB_ERR_TPM, N_("sha1_hash_string: hashing failed"));
	}

	return GRUB_ERR_NONE;
}

/* End TCG Extension */
